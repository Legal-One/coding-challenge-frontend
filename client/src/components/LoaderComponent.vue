<template>
     <svg class="animate-spin h-5 w-5 mr-3" viewBox="0 0 24 24">
          <circle class="path" cx="25" cy="25" r="20" fill="rgb(20 184 166)" stroke-width="5"></circle>
        </svg>
</template>