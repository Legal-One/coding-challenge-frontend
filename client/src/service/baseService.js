import { axiosInstance } from "."

export const getDashboardData=async ()=>{
    let response=await axiosInstance.get('/')
    return response;
}