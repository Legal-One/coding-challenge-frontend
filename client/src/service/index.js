import Axios from 'axios';
const axiosInstance=Axios.create({
    baseURL:'http://localhost:6001/api'
})
export {axiosInstance}