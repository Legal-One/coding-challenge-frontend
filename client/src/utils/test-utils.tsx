import React, { ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import callCenterReducer from '../features/callcenter/callcenter-slice'

function render(
  ui: JSX.Element,
  {
    initialState = {},
    store = configureStore({
      reducer: {
        callCenter: callCenterReducer
      },
      preloadedState: initialState
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }: { children: JSX.Element }) {
    const history = createMemoryHistory()
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    )
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

export * from '@testing-library/react'
export { render }
