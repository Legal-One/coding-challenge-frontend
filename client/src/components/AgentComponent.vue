<template>
    <div class="pl-2">
     <div v-if="loading" class="flex justify-center items-center"><LoaderComponent /></div>
     <TableComponent v-else :columns="['Phone Number','Call Date and time','Resolution']" :list="list" :columnList="['number','dateTime','resolution.resolution']"/>
    </div>
</template>

<script>
import { getAgentData } from '../service/baseService';
import LoaderComponent from './LoaderComponent.vue';
import TableComponent from './TableComponent.vue';

 export default {
    data() {
        return {
            loading: true,
            list: null
        };
    },
    created() {
        // watch the params of the route to fetch the data again
        this.$watch(() => this.$route.params, () => {
            this.fetchData();
        }, 
        // fetch the data when the view is created and the data is
        // already being observed
        { immediate: true });
    },
    methods: {
        async fetchData() {
            this.error = this.post = null;
            this.loading = true;
            try {
                let list = await getAgentData(this.$route.params.id);
                this.list = list.data?.payload?.map((data)=>{
                    let str=data.dateTime.split('T')
                    let time=new Date(data.dateTime).toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true });
                    return {...data,dateTime:str[0]+' '+time}
                });
                this.loading=false;
            }
            catch (error) {
                this.loading=false;
                this.error = error.toString();
            }
        },
    },
    components: { LoaderComponent, TableComponent }
}

</script>

<style lang="scss" scoped>
 .pl-2{
     padding: 2em;
 }
</style>