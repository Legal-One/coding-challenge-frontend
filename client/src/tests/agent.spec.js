import { mount, shallowMount } from "@vue/test-utils";
import AgentComponent from "../components/AgentComponent.vue";
import LoaderComponent from "../components/LoaderComponent.vue";
import * as baseService from "../service/baseService";
jest.mock('../service/baseService')
const mockAxios = jest.genMockFromModule('axios')

baseService.getAgentData = jest.fn().mockImplementation(() => {
    return Promise.resolve({ payload: mockedData })
});

mockAxios.create = jest.fn(() => mockAxios)

 export {mockAxios}

let wrapper;

beforeEach(() => {
     wrapper = mount(AgentComponent,{
        global: {
        mocks: {
            $route: {
              params: {
                id: 'f53b3e0e-6a21-11eb-9439-0242ac130002'
              }
            }
          }
        }
        })
});
const mockedData=[
    {
        "identifier": "f53b3e0e-6a21-11eb-9439-0242ac130002",
        "agentIdentifier": "356b03dc-9ec5-11e7-97a6-d501104f897e",
        "number": "+49151484522",
        "dateTime": "2020-10-05T14:48:00.000Z",
        "duration": 230,
        "resolution": {
            "identifier": "f53b3e0e-6a21-11eb-9439-0242ac130002",
            "resolution": "need reschedule"
        }
    },
    {
        "identifier": "0b96031e-6a22-11eb-9439-0242ac130002",
        "agentIdentifier": "356b03dc-9ec5-11e7-97a6-d501104f897e",
        "number": "+49151484522",
        "dateTime": "2020-10-06T13:50:00.000Z",
        "duration": 93,
        "resolution": {
            "identifier": "0b96031e-6a22-11eb-9439-0242ac130002",
            "resolution": "interested"
        }
    }]

mockAxios.get.mockImplementationOnce(() =>
Promise.resolve({ payload: mockedData }),
)

it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
})

it('should show loader', () => {
    const wrapper = mount(AgentComponent, {
        global: {
            mocks: {
                $route: {
                  params: {
                    id: 'f53b3e0e-6a21-11eb-9439-0242ac130002'
                  }
                }
              }
            },
        data() {
          return {
            loading: true
          }
        }
      })

    expect(wrapper.findComponent(LoaderComponent).isVisible()).toBe(true)
})

it('get api call on load of component', async () => {
    expect(baseService.getAgentData).toHaveBeenCalled()
    // expect(mockAxios.get).toHaveBeenCalledWith('/api/agent/:id')
})