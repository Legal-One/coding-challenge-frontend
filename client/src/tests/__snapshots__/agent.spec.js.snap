// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="pl-2"
>
  <div
    class="flex flex-col"
  >
    <div
      class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
    >
      <div
        class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
      >
        <div
          class="shadow overflow-hidden border-b border-gray-200 sm:rounded-lg"
        >
          <table
            class="min-w-full divide-y divide-gray-200"
          >
            <thead
              class="bg-gray-50"
            >
              <tr>
                
                <th
                  class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                  scope="col"
                >
                  Phone Number
                </th>
                <th
                  class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                  scope="col"
                >
                  Call Date and time
                </th>
                <th
                  class="px-6 py-3 text-left text-xs font-medium text-gray-500 uppercase tracking-wider"
                  scope="col"
                >
                  Resolution
                </th>
                
              </tr>
            </thead>
            <tbody
              class="bg-white divide-y divide-gray-200"
            >
              
              
              <!-- More people... -->
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
